static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,409 @@
+ *
+ * Note that this only applies to specific resize programmatic messages
+ * (these include specific width and height values)
+ */
+const toStandardMessage = (
+	payload: ProgrammaticMessage,
+): StandardMessage<{ width: number; height: number }> => ({
+	id: 'aaaa0000-bb11-cc22-dd33-eeeeee444444',
+	type: 'resize',
+	iframeId: payload.value.id,
+	slotId: payload.value.slotId,
+	value: {
+		height: +payload.value.height,
+		width: +payload.value.width,
+	},
+});
+
+/**
+ * Retrieve a reference to the calling iframe via it's ID.
+ *
+ * Incoming messages contain the ID of the iframe into which the source window is embedded.
+ */
+const getIframe = (data: StandardMessage): HTMLIFrameElement | undefined => {
+	if (data.slotId) {
+		const container = document.getElementById(`dfp-ad--${data.slotId}`);
+		const iframes = container
+			? container.getElementsByTagName('iframe')
+			: null;
+		return iframes?.length ? iframes[0] : undefined;
+	} else if (data.iframeId) {
+		const iframe = document.querySelector<HTMLIFrameElement>(
+			`iframe#${data.iframeId}`,
+		);
+		return iframe ?? undefined;
+	}
+};
+
+// Regex for testing validity of message ids
+const validMessageRegex = /^[a-f0-9]{8}-([a-f0-9]{4}-){3}[a-f0-9]{12}$/;
+
+/**
+ * Narrow an `unknown` payload to the standard message format
+ *
+ * Until DFP provides a way for us to identify with 100% certainty our
+ * in-house creatives, we are left with doing some basic tests
+ * such as validating the anatomy of the payload and whitelisting
+ * event type
+ */
+const isValidPayload = (payload: unknown): payload is StandardMessage => {
+	const payloadToCheck = payload as StandardMessage;
+	return (
+		'type' in payloadToCheck &&
+		'value' in payloadToCheck &&
+		'id' in payloadToCheck &&
+		payloadToCheck.type in LISTENERS &&
+		validMessageRegex.test(payloadToCheck.id)
+	);
+};
+
+/**
+ * Cheap string formatting function
+ *
+ * @param error An object `{ code, message }`. `message` is a string where successive
+ * occurrences of %% will be replaced by the following arguments
+ * @param args Arguments that will replace %%
+ *
+ * @example
+ * formatError({ message: "%%, you are so %%" }, "Regis", "lovely")
+ * => { message: "Regis, you are so lovely" }
+ */
+const formatError = (
+	error: { code: number; message: string },
+	...args: string[]
+) =>
+	args.reduce((e, arg) => {
+		e.message = e.message.replace('%%', arg);
+		return e;
+	}, error);
+
+/**
+ * Convert a posted message to our StandardMessage format
+ *
+ * @param event The message event received on the window
+ * @returns A message with the `StandardMessage` format, or null if the conversion was unsuccessful
+ */
+const eventToStandardMessage = (event: MessageEvent) => {
+	try {
+		// Currently all non-string messages are discarded here since parsing throws an error
+		// TODO Review whether this is the desired outcome
+		const data: unknown = JSON.parse(event.data);
+		return isProgrammaticMessage(data)
+			? toStandardMessage(data)
```

Comment:
I'm not sure if / how much this matters in practice but isn't this a change in the logic of the original js?
In the original, we checked [`isValidPayload`](https://github.com/guardian/frontend/pull/24553/files#diff-d01441202cd6b4393db21ef2274800960087affc6cf6a68f33ec3bfde3af7cacL92) regardless of whether or not `data` was a programmatic message, whereas now we only do that check if it's not a programmatic message.
It seems like [this check](https://github.com/guardian/frontend/pull/24553/files#diff-4a626d643ca6ac867c92cb8cba5d76683fdf05fa87462c5ba844d6fb97c5b506R196) is applicable to both programmatic and non-programmatic.

Reply:
This is a good point! I've reverted to a logic where all kinds of messages are validated. 

I think in the future we shouldn't need to validate a standard message that's been converted from programmatic, since we should trust in the conversion, but it's possible you get different results based on the registered listeners (since it checks `LISTENERS[type]`).

So best to go with the safest option!